design/rich-authorization-requests.md

```diff
@@ -0,0 +1,173 @@
+
+- **authorization_details_types**
+
+Clients announce the authorization data types they use in the new dynamic client registration parameter "authorization_details_types".
+
+
+### Authorization Request
+
+As per spec,
+The `authorization_details` request parameter CAN be used to specify
+authorization requirements in all places where the "scope" parameter
+is used for the same purpose, examples include:
+
+*  Authorization requests as specified in [RFC6749],
+
+*  Access token requests as specified in [RFC6749], if also used as
+   authorization requests, e.g. in the case of assertion grant types
+   [RFC7521],
+
+*  Request objects as specified in [I-D.ietf-oauth-jwsreq],
+
+*  Device Authorization Request as specified in [RFC8628],
+
+*  Backchannel Authentication Requests as defined in [OpenID.CIBA].
+
+we will use it first in Authorization requests.
+
+### Token Response
+
+In addition to the token response parameters as defined in [RFC6749],
+the authorization server MUST also return the authorization details
+as granted by the resource owner and assigned to the respective
+access token.
+
+## Implementation details
+
+SPI
+
+````
+public interface RarProcessorProvider extends Provider {}
+
+public interface RarProcessorProviderFactory extends ProviderFactory<RarProcessorProvider> { }
+````
+e.g. implementation in extension
+
+````
+public class Psd2RarProcessorProvider implements RarProcessorProvider {}
+
+public class Psd2RarProcessorProviderFactory implements RarProcessorProviderFactory { 
+
+    public static final String ID = "rar-psd2";
+    public static final String SUPPORTED_DATA_TYPE = []; //Json array schema definition
+}
+````
+
+* Support advertisement of supported authorization details types in OAuth server metadata
+
+````
+OIDCConfigurationRepresentation config = new OIDCConfigurationRepresentation();
+config.setAuthorizationDataTypesSupported(
+    getSupportedAuthorizationDataTypes(RarProcessorProvider.class)
+    
+);
+````
+
+*  Check if authorization data type or authorization details is not conforming to the respective type
+   definition when authorization_details is available in the authorization requests 
+
+````
+rarProcessorProvider.isAuthorizationDetailsCompliant(authorizationDetails)
+````
+After the user authenticate, rarProcessor should:
+
+*  Determine how authorization details are shown to the user in the user consent
+*  Exchange with the resource server if necessary
+*  Enrich authorization details in the user consent process (e.g. add selected accounts or set expirations)
+*  show the consent page
+
+````
+rarProcessorProvider.prepareAuthorizationDetails()
+````
+   
+When the user submits the consent  
+
+*  Merge requested and pre-existing authorization_details (e.g. in case of update)  
+
+* store authorization_details within the grant
+
+````
+rarProcessorProvider.processAuthorizationDetails()
+````
+
+### Admin UI
+
+AS should allow the possibility to select RarProcessor Implementation Module.
```

Comment:
Should the selection of the RarProcessor implementation / strategy be based only on Realm or also on Client level? 
I can image that different clients might require different details processing strategies, then it would make sense to have be able to override a realm global configuration on client level.

Reply:
I see. For a more detailed configuration, being able to decide the processing strategies at the client level will be a good thing.
 
I would suggest making the first version at the Realm level and then the other version can be extended with client level.

Reply:
This is the question... I was thinking if it is sufficient to stick with the server level for choosing the processing strategy?

Or do you have some use-cases in mind when different clients (or different realms) may want to stick with the different processing requirements?

IMO the `authorization_details` is more advanced version of the `scope` parameter with more fine-grained support. And for example for `scope` parameter, I did not see use-case when different clients want same scope with different meaning. For example that client1 in theory may want to use `scope=profile` to have a bit different meaning than client2 using `scope=profile` . In theory, client1 may want a bit different data to be shown for the `profile` than client2. However in reality, this was never needed...

ATM it seems to me that for choosing the processor, we may also stick to the server level (or provider level) as I pointed in the other place and add more fine-grained control if there is later the use-case for it?

Reply:
One thing, which seems to me will be good for admin UI, might be to being able to choose supported types of authorization_detail types for various clients? This is supported in the client registration parameter `authorization_details_types` and hence probably also need some control in the admin console.

I can imagine also some support for evaluating access tokens (The stuff when you go to client -> Tab `Client Scopes` -> click to some client scope -> tab `Evaluate` to generate example tokens. It can be interesting to be able to generate tokens for various provided `authorization_details` ). However this support for evaluating example token is something to be done as a follow-up PR if needed (Not necessarily in the first PR of this feature IMO).

Reply:
Hi @mposolda , as mentioned in the previous comment processor will be selected at the server level. And we will add the possibility in the admin UI to select the authorization_detail types for a client.